File: JayKay24/Searching 2/iterative-binary-search.ts
/**
 * Returns idx of target, -1 if not found
 * @param nums - a sorted array of integers
 * @param target - integer to search for
 * @returns idx of target, -1 if not found
 */
function iterativeBinarySearch(nums: number[], target: number): number {
  let leftIdx = 0,
    rightIdx = nums.length - 1;

  while (leftIdx <= rightIdx) {
    const midIdx = Math.floor((leftIdx + rightIdx) / 2);

    if (nums[midIdx] === target) return midIdx;

    if (nums[midIdx] < target) {
      leftIdx = midIdx + 1;
    } else {
      rightIdx = midIdx - 1;
    }
  }

  return -1;
}

// Time - O(log(n))
// Space - O(1)

const numbers: number[] = [1, 3, 4, 5, 10, 11, 23, 50];

File: Shizler/Binary-Search/iterativeBinarySearch.py
def iterativeBinarySearch(nums,target):
    rightIdx = len(nums) - 1
    leftIdx = 0
    while leftIdx <= rightIdx:
        midIdx = ((leftIdx + rightIdx) // 2)

        if nums[midIdx] == target: return target
        if target < nums[midIdx]:
            rightIdx = midIdx - 1
        else:
            leftIdx = midIdx + 1
        
    return -1
nums= [1, 3, 4, 5, 10, 11, 23, 50]

print(iterativeBinarySearch(nums,10))